File: cpp/234_PalindromeLinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if(head == NULL || head->next ==NULL) return true;
        //find middle
        ListNode *slow = head;
        ListNode *fast = head;
        while(fast->next != NULL && fast->next->next != NULL){
            slow = slow->next;
            fast = fast->next->next;
        }
        //reverse the second half
        slow->next = reverseList(slow->next);
        slow = slow->next;
        //check whether same
        while(slow != NULL){
            if(head->val != slow->val) return false;
            head = head->next;
            slow = slow->next;
        }
        return true;
    }
private:
    ListNode* reverseList(ListNode* head){
        if(head == NULL || head->next == NULL) return head;
        ListNode* p = head;
        ListNode* pre = NULL;
        while( p != NULL ){
            ListNode *t = p->next;
            p->next = pre;
            pre = p;
            p = t;
        }
        return pre;
    }
};

File: js/234_PalindromeLinkedList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var isPalindrome = function(head) {
    if(head === null || head.next === null) return true;
    //find the middle
    var slow = head;
    var fast = head;
    while(fast.next !== null && fast.next.next !== null){
        slow = slow.next;
        fast = fast.next.next;
    }
    //reverse the second half
    slow.next = reverseList(slow.next);
    slow = slow.next;
    //check whether the same
    while(slow !== null){
        if(head.val !== slow.val) return false;
        head = head.next;
        slow = slow.next;
    }
    return true;
};

var reverseList = function(head){
    if(head === null || head.next === null) return head;
    var pre = null;
    while(head !== null){
        var t = head.next;
        head.next = pre;
        pre = head;
        head = t;
    }
    return pre;
}